### PR TITLE
NIAD-343: Error Handling

### PR DESCRIPTION
**Pull request by Broom94**
Soap error response created, additional error handling for internal errors, extract address from report element, additional error handling for missing soap headers and body elements

### CODE REVIEW

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/utils/ReportParserUtil.java [hunk 1]
```diff
@@ -13,35 +13,63 @@
...
         reportElementsMap.put(ReportElement.TRACKING_ID, getTrackingId(document));
         return reportElementsMap;
     }
 
-    private static String getMessageId(Document document) {
+    private static String getMessageId(Document document) throws SoapClientException {
+        String messageId = "MessageId";
+        String errorLocation = "Header";
         Node messageIdNode = document.selectSingleNode(MESSAGE_ID_NODE);
+        try {
```
Comment (dawidskainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/utils/ReportParserUtil.java:38]: Let's not catch NPE. Let's throw SoapClientException if messageIdNode is null

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/utils/CustomXMLException.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package uk.nhs.adaptors.oneoneone.cda.report.controller.utils;
```
Comment (dawidskainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/utils/CustomXMLException.java:1]: utils is not the best package for this class
Maybe let's create a dedicated exceptions package?

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/utils/CustomXMLException.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+import lombok.Getter;
+
+@Getter
+public class CustomXMLException extends Exception {
```
Comment (dawidskainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/utils/CustomXMLException.java:6]: In don't like 'Custom' prefix. Maybe ItkXmlException? Or CdaXmlException?

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/utils/ReportParserUtil.java [hunk 4]
```diff
@@ -13,35 +13,63 @@
...
+            return messageIdNode.getText();
+        } catch (NullPointerException e) {
+            throw new SoapClientException(messageId, errorLocation, e);
+        }
+    }
+
+    private static String getAddress(Document document) {
+        String defaultToAddress = "http://www.w3.org/2005/08/addressing/anonymous";
```
Comment (dawidskainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/utils/ReportParserUtil.java:46]: It's not this class's responsibility to know Error response default reply to address. This class should only know how to extract Address from SOAP.

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java [hunk 5]
```diff
@@ -46,23 +55,58 @@
...
-        } catch (XmlException | DocumentException e) {
+        } catch (DocumentException e) {
+            LOGGER.error(BAD_REQUEST.toString() + e.getMessage());
+            throw new ResponseStatusException(BAD_REQUEST, createErrorResponse(
+                toAddress, messageId, CLIENT_ERROR_CODE, "This is not a valid XML message", e.getMessage()));
+        } catch (XmlException e) {
+            LOGGER.error(BAD_REQUEST.toString() + e.getMessage());
+            throw new ResponseStatusException(BAD_REQUEST, createErrorResponse(
+                toAddress, messageId, CLIENT_ERROR_CODE, "Message body not valid", e.getMessage()));
+        } catch (ItkXmlException e) {
```
Comment (dawidr-kainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java:86]: `catch (ItkXmlException | SoapClientException e)` could avoid code duplications. ;)

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java [hunk 6]
```diff
@@ -46,23 +55,58 @@
         produces = TEXT_XML_VALUE)
     @ResponseStatus(value = ACCEPTED)
     public ResponseEntity<String> postReport(@RequestBody String reportXml) {
+        String toAddress = null;
```
Comment (dawidr-kainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java:58]: ` = null;` is not needed I believe. :P

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java (hunk 5 above)
Comment (Broom94) [reply service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java:86]: Cannot resolve method 'getReason' in 'Exception'

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java (hunk 6 above)
Comment (Broom94) [reply service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java:58]: Is needed for Exception catch

Comment (dawidr-kainos) [reply service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java:58]: I mean `null` is default value for string anyway. ;)

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java (hunk 5 above)
Comment (dawidr-kainos) [reply service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java:86]: Interesting. What I see both exceptions have this method O.o

**Pull request merged**